core/src/services/s3/backend.rs

```diff
@@ -517,6 +518,60 @@ impl S3Builder {
 
         self
     }
+
+    /// a helper function to make it easier to find region
+    pub async fn detect_region(&self) -> Result<Option<String>> {
+        if let Some(region) = &self.region {
+            return Ok(Some(region.to_string()));
+        }
+
+        let mut endpoint = match &self.endpoint { 
+            Some(endpoint) => { 
+                if endpoint.starts_with("http") { 
+                    endpoint.to_string() 
+                } else { 
+                    // Prefix https if endpoint doesn't start with scheme. 
+                    format!("https://{}", endpoint) 
+                } 
+            } 
+            None => "https://s3.amazonaws.com".to_string(), 
+        }; 
+     
+        // If endpoint contains bucket name, we should trim them. 
+        endpoint = endpoint.replace(&format!("//{0}.", &self.bucket), "//");
+
+        let url = format!("{0}/{1}", endpoint, &self.bucket); 
+
+        debug!("backend detect region with url: {url}"); 
+     
+        let req = http::Request::head(&url).body(AsyncBody::Empty).map_err(new_request_build_error)?; 
+     
+        let client = HttpClient::new()?;
+        let res = client.send(req).await?; 
+     
+        debug!( 
+            "auto detect region got response: status {:?}, header: {:?}", 
+            res.status(), 
+            res.headers() 
+        ); 
+        
+        match res.status() { 
+            // The endpoint works, return with not changed endpoint and 
+            // default region. 
+            StatusCode::OK | StatusCode::FORBIDDEN | StatusCode::METHOD_NOT_ALLOWED => { 
+                let region = res 
+                    .headers() 
+                    .get("x-amz-bucket-region") 
+                    .unwrap_or(&HeaderValue::from_static("us-east-1")) 
+                    .to_str() 
+                    .map_err(|e| println!("{e:?}")).unwrap() 
+                    .to_string(); 
+                Ok(Some(region)) 
+            }
+            // Unexpected status code 
+            code => Err(Error::new(ErrorKind::NotFound, format!("can't detect region automatically, no valid endpoint template for {endpoint} {code:?}").as_str()))
```

Comment:
If we failed to find the region (no matter why), we should better return `None` instead.